File: src/cppcache/src/CqEventImpl.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "CqEventImpl.hpp"
#include <gfcpp/CacheableString.hpp>
#include "ThinClientPoolHADM.hpp"
#include "ThinClientCacheDistributionManager.hpp"
#include "TcrMessage.hpp"

using namespace apache::geode::client;
CqEventImpl::CqEventImpl(CqQueryPtr& cQuery,
                         CqOperation::CqOperationType baseOp,
                         CqOperation::CqOperationType cqOp,
                         CacheableKeyPtr& key, CacheablePtr& value,
                         ThinClientBaseDM* tcrdm, CacheableBytesPtr deltaBytes,
                         EventIdPtr eventId)
    : m_error(false) {
  m_cQuery = cQuery;
  m_queryOp = cqOp;
  m_baseOp = baseOp;
  m_key = key;
  m_newValue = value;
  if (m_queryOp == CqOperation::OP_TYPE_INVALID) m_error = true;
  m_tcrdm = tcrdm;
  m_deltaValue = deltaBytes;
  m_eventId = eventId;
}

CqQueryPtr CqEventImpl::getCq() const { return m_cQuery; }

CqOperation::CqOperationType CqEventImpl::getBaseOperation() const {
  return m_baseOp;
}

/**
 * Get the the operation on the query results. Supported operations include
 * update, create, and destroy.
 */
CqOperation::CqOperationType CqEventImpl::getQueryOperation() const {
  return m_queryOp;
}

/**
 * Get the key relating to the event.
 * @return Object key.
 */
CacheableKeyPtr CqEventImpl::getKey() const { return m_key; }
/**
 * Get the new value of the modification.
 *  If there is no new value because this is a delete, then
 *  return null.
 */
CacheablePtr CqEventImpl::getNewValue() const {
  if (m_deltaValue == NULLPTR) {
    return m_newValue;
  } else {
    // Get full object for delta
    TcrMessageRequestEventValue fullObjectMsg(m_eventId);
    TcrMessageReply reply(true, NULL);
    ThinClientPoolHADM* poolHADM = dynamic_cast<ThinClientPoolHADM*>(m_tcrdm);
    GfErrType err = GF_NOTCON;
    if (poolHADM) {
      err = poolHADM->sendRequestToPrimary(fullObjectMsg, reply);
    } else {
      err = static_cast<ThinClientCacheDistributionManager*>(m_tcrdm)
                ->sendRequestToPrimary(fullObjectMsg, reply);
    }
    CacheablePtr fullObject = NULLPTR;
    if (err == GF_NOERR) {
      fullObject = reply.getValue();
    }
    return fullObject;
  }
}

bool CqEventImpl::getError() { return m_error; }

std::string CqEventImpl::toString() {
  char buffer[1024];
  ACE_OS::snprintf(
      buffer, 1024,
      "CqEvent CqName=%s; base operation=%d; cq operation= %d;key=%s;value=%s",
      m_cQuery->getName(), m_baseOp, m_queryOp, m_key->toString()->asChar(),
      m_newValue->toString()->asChar());
  return buffer;
}

CacheableBytesPtr CqEventImpl::getDeltaValue() const { return m_deltaValue; }

File: geode-cq/src/main/java/org/apache/geode/cache/query/cq/internal/CqEventImpl.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
 * agreements. See the NOTICE file distributed with this work for additional information regarding
 * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance with the License. You may obtain a
 * copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package org.apache.geode.cache.query.cq.internal;


/**
 * Interface for CqEvent. Offers methods to get information from CqEvent.
 *
 * @since GemFire 5.5
 */

import org.apache.geode.InvalidDeltaException;
import org.apache.geode.cache.Operation;
import org.apache.geode.cache.client.internal.QueueManager;
import org.apache.geode.cache.query.CqEvent;
import org.apache.geode.cache.query.CqQuery;
import org.apache.geode.internal.cache.EventID;

public class CqEventImpl implements CqEvent {

  private final CqQuery cQuery;
  private final Operation baseOp;
  private final Operation queryOp;
  private final Object newValue;
  private final Object key;
  private final byte[] delta;
  private Throwable throwable = null;
  private final QueueManager qManager;
  private final EventID eventId;

  CqEventImpl(CqQuery cQuery, Operation baseOp, Operation cqOp, Object key, Object value,
      byte[] deltaVal, QueueManager qManager, EventID eventId) {
    this.cQuery = cQuery;
    queryOp = cqOp;
    this.baseOp = baseOp;
    this.key = key;
    newValue = value;
    delta = deltaVal;
    // Handle Exception event.
    if (queryOp == null) {
      setException();
    }
    this.qManager = qManager;
    this.eventId = eventId;
  }

  @Override
  public CqQuery getCq() {
    return cQuery;
  }

  /**
   * Get the operation on the base region that triggered this event.
   */
  @Override
  public Operation getBaseOperation() {
    return baseOp;
  }

  /**
   * Get the the operation on the query results. Supported operations include update, create, and
   * destroy.
   */
  @Override
  public Operation getQueryOperation() {
    return queryOp;
  }

  /**
   * Get the key relating to the event.
   *
   * @return Object key.
   */
  @Override
  public Object getKey() {
    return key;
  }

  /**
   * Get the old value that was modified. If there is no old value because this is an insert, then
   * return null.
   */
  /*
   * public Object getOldValue() { return this.oldValue; }
   */

  /**
   * Get the new value of the modification. If there is no new value because this is a delete, then
   * return null.
   */
  @Override
  public Object getNewValue() {
    if (newValue == null && delta != null) {
      throw new InvalidDeltaException();
    }
    return newValue;
  }

  /**
   * If an error occurred, return the Throwable, otherwise return null. If an error occurred, then
   * this event will be passed to the <code>onError</code> method of the CqListener instead of the
   * <code>onEvent</code> method.
   */
  @Override
  public Throwable getThrowable() {
    return throwable;
  }

  @Override
  public byte[] getDeltaValue() {
    return delta;
  }

  /**
   * Checks if the Cq Event is for an exception. If so constructs the exception.
   */
  public void setException() {
    // Needs to be changed.
    throwable = new Throwable(
        "Exception occurred while applying query on a cache event.");
  }

  public void setException(String exceptionText) {
    throwable = new Throwable(exceptionText);
  }

  public QueueManager getQueueManager() {
    return qManager;
  }

  public EventID getEventID() {
    return eventId;
  }

  @Override
  public String toString() {
    StringBuffer buffer = new StringBuffer();
    buffer.append("CqEvent [").append("CqName=").append(cQuery.getName())
        .append("; base operation=").append(baseOp).append("; cq operation=")
        .append(queryOp).append("; key=").append(key).append("; value=")
        .append(newValue).append("]");
    return buffer.toString();
  }
}
